hs628_tools/regression_check.py

```diff
@@ -0,0 +1,200 @@
+        https://github.com/dziganto/Data_Science_Fundamentals/blob/master/notebooks/Machine_Learning/
+        Supervised_Learning/Regression/Linear_Regression/3_Linear_Regression_Assumptions_and_Evaluation.ipynb
+
+        The following functions are defined in this class :
+
+         func : check_linearity_residual_homoscedasticity_independence_plt checks to see if the
+                data can fit to a linear model/non-linear model. Also check the homoscedasticity
+                (check for the spread of data) of the residuals/error terms and also check for their
+                independence. We diagnose these through a Residuals vs Predicted values plot.If there
+                is no clear pattern and data symmetrically distributed it can be said to satisfy the
+                assumptions.
+
+         func : check_residuals_normality_plt checks to see the normality of the error terms through
+                a density plot of the residuals. If the graph looks like gaussian normal curve it can
+                be said that the residuals satisfy the assumption of normality.
+
+         func : check_predictors_independence_plt checks for the independence of each predictor by
+                plotting a heatmap of the correlations of each predictor with one and another. The
+                heatmap would clearly indicate if there are multicolinear predictors.
+
+         func : regression_metric builds a metric for evaluating the model. It is a harmonic mean of
+                the R-Squared value and the F-Statistic taking into consideration the probability of
+                significance of the F-statistic being significant than having an only intercept model.
+    """
+
+    def __init__(self):
+        # Instantiating the parameters used in the methods below :PredictorColumnnames,R2,
+        # F-stat,F-statpval,ResidualSumofSquares,RegressionMetric
+        self.colnames = None
+        self.rsquared = 0
+        self.fstatistic = 0
+        self.fstat_pval = 0
+        self.resid_sumsq = 0
+        self.regmetric = 0
+
+    def check_linearity_residual_homoscedasticity_independence_plt(self, X, y, model):
+        """
+        Plots the data fitted to a linear model/non-linear model and helps user in checking the homoscedasticity
+         of the residuals/error terms and also checks for their independence.
+            -----------
+            Param:
+                X: 2-D numpy array of the predictors
+                y: 1-D numpy array of the response variables
+                model : Linear regression model fitted to X(predictors) and y(response)
+            -----------
+            Return:
+                Doesn't return anything, just shows the plot object built
+        """
+
+        # Plot of Residuals vs predicted values from the linear model
+        fig, axes = plt.subplots(1, 1, sharex=False, sharey=False)
+        fig.suptitle('Residual Plots')
+        axes.plot(model.predict(X), y - model.predict(X), 'bo--')
```

Comment:
This is what I was referring to, where we rely on the model having a 'predict' method